README.md
# login-utils

  login-utils is a very simple user authentication library. The library has a default User MongoDB model (using Mongoose).
Moreover, the encryption of the password is performed using bcrypt.

## Installation

    $ npm install login-utils
    $ export MONGODB_CONNECT="mongodb://[yourUserName]:[yourPassword]@[your_db_host:your_db_port]/[your_db_name]"

  Setting environment variable MONGODB_CONNECT will enable Mongoose to tap into a remote MongoDB resource ([MongoLab](http://www.mongolab.com) for example). If it is unset, localhost port 27017 is assumed.

## Usage

  The api is very straightforward to use.

### Changing bcrypt number of rounds (default is 11):
```js
  var loginUtils = require("login-utils");
  loginUtils.setBcryptNumberOfRounds(15);

  //You can get the current number of rounds via
  var nbOfRounds = loginUtils.getBcryptNumberOfRounds();
  console.log("Current bcrypt number of rounds is %d", nbOfRounds);
```


### Finding a user by his email address:

```js
  var loginUtils = require("login-utils");
  var email = "kenshiro@some-email.com";

  loginUtils.findUserByEmail(email, function(err, user){
    if(err){
      console.log("An error occurred while trying to find user by email: %s", err);
    }else{
      console.log("User has been retrieved: [id=%s, first-name=%s, last-name=%s, user-name=%s" +
                  ", email=%s, hashed-password=%s, bcrypt-salt=%s]", user._id, user.firstName, user.lastName,
                  user.userName, user.email, user.password, user.salt);
    }
  });

```

### Finding a user by his/her user name:

```js
  var loginUtils = require("login-utils");
  var userName = "kenshiro.hackuto";

  loginUtils.findUserByUserName(userName, function(err, user){
    if(err){
      console.log("An error occurred while trying to find user by user name: %s", err);
    }else{
      console.log("User has been retrieved: [id=%s, first-name=%s, last-name=%s, user-name=%s" +
                  ", email=%s, hashed-password=%s, bcrypt-salt=%s]", user._id, user.firstName, user.lastName,
                  user.userName, user.email, user.password, user.salt);
    }
  });

```


### Finding a user by his/her first and last name:

```js
  var loginUtils = require("login-utils");
  var firstName = "kenshiro";
  var lastName = "hackuto";

  loginUtils.findUsersByFirstAndLastName(firstName, lastName, function(err, user){
    if(err){
      console.log("An error occurred while trying to find user by first and last name: %s", err);
    }else{
      console.log("User has been retrieved: [id=%s, first-name=%s, last-name=%s, user-name=%s" +
                  ", email=%s, hashed-password=%s, bcrypt-salt=%s]", user._id, user.firstName, user.lastName,
                  user.userName, user.email, user.password, user.salt);
    }
  });

```




### Checking user login details against what is stored

```js
   var loginUtils = require("login-utils");
   var password = "Hackuto-Shinken is invincible";
   var email = "kenshiro@hackuto-shinken-mail.com";

   loginUtils.loginCheck(email, password, function(err, response){
     if(err){
      console.log("An error occurred when authenticating user: %s", err);
     }else if(!response.user){
      console.log(response.message);
     }else{
      var user = response.user;
      console.log("Successfully authenticated user: [id=%s, first-name=%s, last-name=%s, user-name=%s" +
                        ", email=%s, hashed-password=%s, bcrypt-salt=%s]", user._id, user.firstName, user.lastName, user.userName,
                        user.email, user.password, user.salt);
     }
   });

```

### Creating a user

```js
  var loginUtils = require("login-utils");
  var firstName = "kenshiro";
  var lastName = "hackuto";
  var password = "Hackuto-Shinken is invincible";
  var email = "kenshiro@hackuto-shinken-mail.com";

  loginUtils.createNewUser({
      firstName: firstName,
      lastName: lastName,
      userName: firstName + "." + lastName,
      password: password,
      email: email
    }, function(err, user){
      if(err){
        console.log("An error occurred creating user: %s", err);
      }else{
        console.log("Successfully created user: [id=%s, first-name=%s, last-name=%s, user-name=%s" +
                    ", email=%s, hashed-password=%s, bcrypt-salt=%s]", user._id, user.firstName, user.lastName, user.userName
                    user.email, user.password, user.salt);
      }
    });
```

### Creating a social media user
Note that this will also create a user, if not user with a matching email address is found.

```js
    var loginUtils = require("login-utils");
    var firstName = "kenshiro";
    var lastName = "hackuto";
    var password = "Hackuto-Shinken is invincible";
    var email = "kenshiro@hackuto-shinken-mail.com";
    var providerName = "Facebook";
    var providerUserId = "1234Fbk07";
    var displayName = "kenshiro-o";

    var socialMediaUserData = {
      firstName: firstName,
      lastName: lastName,
      userName: kenshiro-o,
      email: email,
      providerName: providerName,
      providerUserId: providerUserId,
      displayName: displayName;
    };

    loginUtils.createSocialMediaUser(socialMediaUserData, function(err, socialMediaUser){
      if(err){
          console.log("An error occurred when trying to create social media user [error=%s]", err);
      }else{
        var user = socialMediaUser.user;
        console.log("Successfully created a social media user [social-media-user-id=%s, provider-name=%s, provider-user-id=%s, display-name=%s]",
                     socialMediaUser._id, socialMediaUser.providerName, socialMediaUser.providerUserId, socialMediaUser.displayName);

        console.log("User associated with social media user is the following: [user-id=%s, first-name=%s, last-name=%s," +
                    "email=%s, hashed-password=%s, bcrypt-salt=%s]", user._id, user.firstName, user.lastName, user.email,
                    user.password, user.salt);

      }
    });

```

### Create a password reset token

```js
   var loginUtils = require("login-utils");
   var email = "kenshiro@hackuto-shinken-mail.com";
   //Token expires in 5 hours
   var expiryInMillis = 1000 * 60 * 60 * 5;
   loginUtils.setPasswordResetTokenValidityInMillis(expiryInMillis);

   loginUtils.createPasswordResetToken(email, function(err, token){
     if(err){
      console.log("An error occurred when trying to create password reset token for user: %s", err);
     } else{
        console.log("Successfully created a password reset token [user-id=%s, token-string=%s, token-expiry=%s]",
                    token.user, token.token, token.expiry);
     }
   });

```

### Find a password reset token by token string

```js
   var loginUtils = require("login-utils");
   var tokenString = "GIBBERISH_)821832jkhhfdshdfhdls";

   loginUtils.findPasswordResetTokenByTokenString(tokenString, function(err, token){
     if(err){
      console.log("An error occurred when trying to find password reset token by token string: %s", err);
     }else if(!token){
      console.log("No such password reset token exists [token-string=%s]", tokenString);
     }else{
        console.log("Found password reset token [user-id=%s, token-string=%s, token-expiry=%s]",
                    token.user, token.token, token.expiry);
     }
   });

```







## Additional features
  - Retrieving user by id
  - customizing bcrypt encrypter (e.g. configuring the number of rounds)
  - Remote login support provided by [Austin Mullins](http://github.com/amullins83)

## Licence

(The MIT License)

Copyright (c) 2013 Kenshiro &lt;kenshiro@kenshiro.me&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
0.1.0 / 2013-06-19
====================
* Refactored for better remote support

0.0.7 / 2013-04-06
====================
* Enable setting of bcrypt number of rounds
* Strengthen hash comparison against timing attacks

0.0.6 / 2013-03-24
====================
* Added password reset token:
  - can create one from a given existing user's email address
  - can set the expiry of the token (in milliseconds)
  - can find one by token string

0.0.5 / 2013-03-24
====================
* Enable searching of users by first and last name
* Improved documentation

0.0.4 / 2013-03-24
====================
* Added userName to User model object. It is a unique field!
* Enabled searching by userName

0.0.3 / 2013-03-17
====================
* Added SocialMediaUser integration:
  - A user can now be associated with multiple social user personae
  - Moreover the social media user has a reference to the user
  - Increased test coverage


0.0.2 / 2013-03-06
====================
* Added git repo and keywords information to package.json

0.0.1 / 2013-03-05
====================

* Initial release
* Basic code for client authentication

test/loginCheckerTest.js
var vows = require("vows"),
  mongoose = require("mongoose"),
  assert = require("assert"),
  util = require("util"),
  UserSchema = require("./../lib/model/UserSchema"),
  SocialMediaUserSchema = require("./../lib/model/SocialMediaUserSchema"),
  loginChecker = require("./../lib/loginChecker"),
  expect = require("expect.js");

//mongoose.connect("mongodb://127.0.0.1/login-utils-test");
var firstName = "kenshiro";
var lastName = "hackuto";
var userName = "kenshiro.hokuto";
var password = "Hackuto-Shinken is invincible";
var email = "kenshiro@hackuto-shinken-mail.com";

var providerName = "Facebook";
var providerUserId = "FacebookUserId-Kenshiro";

var tokenString = "";

var defaultBcryptSaltFactor = 11;

var nullFunction = function() {};

vows.describe("User checks").addBatch({
  "Delete all collections":{
    topic: function(){
      for (var i in mongoose.connection.collections) {
        mongoose.connection.collections[i].remove(nullFunction);
      }
      this.callback(null);
    },

    "Collections are deleted": function(err){
      expect(err).to.be(undefined);
    }

  }


}).addBatch({
    "When creating a user that does not yet exist": {
      topic: function () {
        loginChecker.createNewUser({
          firstName: firstName,
          lastName: lastName,
          userName: userName,
          password: password,
          email: email
        }, this.callback);
      },

      "A new user is returned": function (err, user) {
        assert.ok(!err);
        assert.ok(user);
        assert.deepEqual(user.firstName, firstName);
        assert.deepEqual(user.lastName, lastName);
        assert.deepEqual(user.userName, userName);
        assert.deepEqual(user.email, email);
        assert.ok(user.password);
        assert.ok(user.salt);
      }
    }
  }).addBatch({
    "When creating a user that already exists": {
      topic: function () {
        loginChecker.createNewUser({
          firstName: firstName,
          lastName: lastName,
          userName: userName,
          password: password,
          email: email
        }, this.callback);
      },

      "An error is returned": function (err, user) {
        assert.ok(err);
      }
    },

    "When searching for a user that does not exist": {
      topic: function () {
        loginChecker.findUserByEmail("kenshiro@email.com", this.callback);
      },

      "No user is returned": function (err, response) {
        assert.ok(!err);
        assert.ok(!response);
      }
    },

    "When searching for a user that does exist": {
      topic: function () {
        loginChecker.findUserByEmail(email, this.callback);
      },

      "The expected user is returned": function (err, user) {
        assert.ok(!err);
        assert.ok(user);
        assert.deepEqual(user.email, email);
        assert.deepEqual(user.firstName, firstName);
        assert.deepEqual(user.lastName, lastName);
        assert.deepEqual(user.userName, userName);
      }
    },

    "When searching for a user name that does not exist": {
      topic: function () {
        loginChecker.findUserByUserName("kenshiro.notExists", this.callback);
      },

      "No user is returned": function (err, response) {
        assert.ok(!err);
        assert.ok(!response);
      }
    },

    "When searching for a user name that does exist": {
      topic: function () {
        loginChecker.findUserByUserName(userName, this.callback);
      },

      "The expected user is returned": function (err, user) {
        assert.ok(!err);
        assert.ok(user);
        assert.deepEqual(user.email, email);
        assert.deepEqual(user.firstName, firstName);
        assert.deepEqual(user.lastName, lastName);
        assert.deepEqual(user.userName, userName);
      }
    },

    "When searching for users that does not exist by first and last name": {
      topic: function () {
        loginChecker.findUsersByFirstAndLastName(firstName + "_O", lastName, this.callback);
      },

      "The expected collection of user is returned": function (err, users) {
        assert.ok(!err);
        assert.deepEqual(users.length, 0);
      }
    },


    "When searching for users that does exist by first and last name": {
      topic: function () {
        loginChecker.findUsersByFirstAndLastName(firstName, lastName, this.callback);
      },

      "The expected collection of user is returned": function (err, users) {
        assert.ok(!err);
        assert.deepEqual(users.length, 1);
        var user = users[0];
        assert.ok(user);
        assert.deepEqual(user.email, email);
        assert.deepEqual(user.firstName, firstName);
        assert.deepEqual(user.lastName, lastName);
        assert.deepEqual(user.userName, userName);
      }
    },




    "When trying to login with an incorrect password": {
      topic: function () {
        loginChecker.loginCheck(email, "123", this.callback);
      },

      "An error is returned": function (err, response) {
        assert.ok(!err);
        assert.ok(!response.user);
        assert.ok(response.message);
      },

      "When trying to login with a correct password": {
        topic: function () {
          loginChecker.loginCheck(email, password, this.callback);
        },

        "A user is returned": function (err, response) {
          assert.ok(!err);
          assert.ok(response.user);
          assert.deepEqual(response.user.email, email);
          assert.deepEqual(response.user.firstName, firstName);
          assert.deepEqual(response.user.lastName, lastName);
        }
      }
    },

    "When creating a social media persona for an already existing user":{
      topic: function(){
        var socialMediaUserData = {
          firstName: firstName,
          lastName: lastName,
          userName: userName,
          email: email,
          providerName: providerName,
          providerUserId: providerUserId,
          displayName: firstName
        };
        loginChecker.createSocialMediaUser(socialMediaUserData, this.callback);
      },

      "A social media user is returned": function(err, socialMediaUser){
        expect(err).to.be(null);
        expect(socialMediaUser.user.email).to.be(email);
        expect(socialMediaUser.providerUserId).to.be(providerUserId);
        expect(socialMediaUser.providerName).to.be(providerName);
        expect(socialMediaUser.displayName).to.be(firstName);
      }
    },

    "When creating a social media persona for a user that does not exist":{
      topic: function(){
        var socialMediaUserData = {
          firstName: firstName,
          lastName: lastName,
          userName: userName + "2",
          email: "kenshiro-test2@mail.com",
          providerName: providerName,
          providerUserId: "12345",
          displayName: "Ken123"
        };
        loginChecker.createSocialMediaUser(socialMediaUserData, this.callback);
      },

      "A social media user is returned": function(err, socialMediaUser){
        expect(err).to.be(null);
        expect(socialMediaUser.user.email).to.be("kenshiro-test2@mail.com");
        expect(socialMediaUser.providerUserId).to.be("12345");
        expect(socialMediaUser.providerName).to.be(providerName);
        expect(socialMediaUser.displayName).to.be("Ken123");
      }
    }

  }).addBatch({
    "When looking for a social media user that has not been created yet": {
      topic: function () {
        loginChecker.findSocialMediaUser("Facebook", "1234fbk", this.callback);
      },

      "A null entry is returned": function (err, socialMediaUser) {
        expect(err).to.be(null);
        expect(socialMediaUser).to.be(null);

      }
    },


    "When creating a password reset token for a user that does not exist":{
      topic: function(){
        loginChecker.createPasswordResetToken("NonExistantMail", this.callback);
      },

      "Nothing is returned":function(err, token){
        expect(err).to.be(null);
        expect(token).to.be(null);
      }
    },

    "When creating a password reset token for a user that does exist":{
      topic: function(){
        loginChecker.createPasswordResetToken(email, this.callback);
      },

      "A valid token is returned":function(err, token){
        expect(err).to.be(null);
        expect(token).to.be.ok();
        tokenString = token.token;
      }
    }

  }).addBatch({

    "When searching for password reset tokens for a user that does exist":{
      topic: function(){
        loginChecker.findPasswordResetTokensForUser(email, this.callback);
      },

      "One token is returned":function(err, tokens){
        expect(err).to.be(null);
        expect(tokens.length).to.be(1);
      }
    },

    "When searching for a password reset token by token":{
      topic: function(){
        loginChecker.findPasswordResetTokenByTokenString(tokenString, this.callback);
      },

      "The expected token is returned ":function(err, token){
        expect(err).to.be(null);
        expect(token).to.be.ok();
      }
    }
  }).addBatch({
    "When setting the bcrypt number of roundsr to an incorrect value":{
      topic:function(){
        var previousValue = loginChecker.getBcryptNumberOfRounds();
        loginChecker.setBcryptNumberOfRounds("EL");
        return previousValue;
      },

      "The previous value is kept": function(previousValue){
        expect(previousValue).to.be(defaultBcryptSaltFactor);
      }
    },

    "When setting the bcrypt number of rounds to a correct value":{
       topic:function(){
         loginChecker.setBcryptNumberOfRounds(100);
         return loginChecker.getBcryptNumberOfRounds();
       },

       "The value is updated": function(numberOfRounds){
         expect(numberOfRounds).to.be(100);
       }
     },

    "When setting the bcrypt number of rounds to a correct then incorrect value":{
      topic:function(){
        loginChecker.setBcryptNumberOfRounds(100);
        loginChecker.setBcryptNumberOfRounds("EL");
        return loginChecker.getBcryptNumberOfRounds();
      },

      "The previous correct value is kept": function(numberOfRounds){
        expect(numberOfRounds).to.be(100);
      }
    }
  }).addBatch({
    "Cleanup": {
      topic: function () {
        return true;
      },

      "Exist gracefully": function (topic) {
        mongoose.connection.close();
        setTimeout(function () {
          process.exit(0);
        }, 2000);

      }
    }
  }).run();

lib/loginChecker.js
(function() {

  "use strict";

  var bcryptUtils = require("./util/bcryptUtils"),
    UserSchema = require("./model/UserSchema"),
    SocialMediaUserSchema = require("./model/SocialMediaUserSchema"),
    PasswordResetTokenSchema = require("./model/PasswordResetTokenSchema"),
    mongoose = require("mongoose"),
    moment = require("moment"),
    crypto = require('crypto'),
    connectString;

  if(process.env.MONGODB_CONNECT && !process.env.TEST_MODE)
    connectString = process.env.MONGODB_CONNECT;
  else
    connectString = "mongodb://127.0.0.1/login-utils-test";

  module.exports.mongoose = mongoose.connect(connectString);

  var User = mongoose.model("User", UserSchema);
  var SocialMediaUser = mongoose.model("SocialMediaUser", SocialMediaUserSchema);
  var PasswordResetToken = mongoose.model("PasswordResetToken", PasswordResetTokenSchema);

  var TOKEN_LENGTH = 48;

  //Default token validity is 1 hour
  var passwordResetTokenValidityInMillis = 1000 * 60 * 60;

  module.exports.setBcryptNumberOfRounds = function(saltFactor){
    bcryptUtils.setBcryptNumberOfRounds(saltFactor);
  };

  module.exports.getBcryptNumberOfRounds = function(){
    return bcryptUtils.getBcryptNumberOfRounds();
  };

  module.exports.setPasswordResetTokenValidityInMillis = function(expiryInMillis){
    var ret = expiryInMillis;
    if(typeof expiryInMillis === "number" && !isNaN(expiryInMillis)){
      passwordResetTokenValidityInMillis = expiryInMillis;
    }else{
      ret = new Error("The argument supplied is not a number");
    }

    return ret;
  };

  module.exports.createPasswordResetToken = function(email, callback){
    findUserByEmail(email, function(err, user){
      if(err){
        process.nextTick(function(){
          callback(err);
        });
      }else if(!user){
        process.nextTick(function(){
          callback(null, null);
        });
      }else{
        var expiryDate = moment().add("ms", passwordResetTokenValidityInMillis).toDate();
        crypto.randomBytes(TOKEN_LENGTH, function(err, buf){
          if(err){
            process.nextTick(function(){
              callback(err);
            });
          }else{
            var tokenString = buf.toString("hex");
            var token = new PasswordResetToken({
              user: user._id,
              token: tokenString,
              expiry: expiryDate
            });
            token.save(function(err){
              process.nextTick(function(){
                callback(err, err ? null : token);
              });
            });
          }
        });
      }
    });
  };

  module.exports.findPasswordResetTokenByTokenString = function(tokenString, callback){
    PasswordResetToken.findOne({token: tokenString}, callback);
  };

  module.exports.findPasswordResetTokensForUser = function(email, callback){
    findUserByEmail(email, function(err, user){
      if(err){
        process.nextTick(function(){
          callback(err);
        });
      }else if(!user){
        process.nextTick(function(){
          callback(null);
        });
      }else{
        PasswordResetToken.find({user: user._id}, callback);
      }
    });
  };

  /**
   * This method assumes the email and password have already been validated AND sanitized
   * @param email the user's email
   * @param password the user's password
   * @param callback
   */
  module.exports.loginCheck = function (email, password, callback) {
    User.findOne({email: email}, function (err, user) {
      if (err) {
        process.nextTick(function () {
          callback(err);
        });
      } else if (!user) {
        process.nextTick(function () {
          callback(null, {user: null, message: "User with email '" + email + "' does not exist in store"});
        });
      } else {
        bcryptUtils.checkHash(password, user.salt, user.password, function (err, match) {
          if (err) {
            process.nextTick(function () {
              callback(err);
            });
          } else {
            if (!match) {
              process.nextTick(function () {
                callback(null, {user: null, message: "Incorrect user/password combination. Please try again."});
              });
            } else {
              process.nextTick(function () {
                callback(null, {user: user});
              });
            }
          }
        });
      }
    });
  };

  function findUserByUserName(userName, callback){
    User.findOne({userName: userName}, callback);
  }

  function findUserByEmail(userEmail, callback) {
    User.findOne({email: userEmail}, callback);
  }

  function findUsersByFirstAndLastName(firstName, lastName, callback){
    User.find({firstName: firstName, lastName: lastName}, callback);
  }

  module.exports.findUserByUserName = findUserByUserName;
  module.exports.findUserByEmail = findUserByEmail;
  module.exports.findUsersByFirstAndLastName = findUsersByFirstAndLastName;

  function findSocialMediaUser (providerName, providerUserId, callback) {
    SocialMediaUser.findOne({providerName: providerName, providerUserId: providerUserId})
      .populate("user")
      .exec(function (err, socialMediaUser) {
        if (err) {
          process.nextTick(function () {
            callback(err);
          });
        } else {
          process.nextTick(function () {
            callback(null, socialMediaUser);
          });
        }
      });
  }

  module.exports.findSocialMediaUser = findSocialMediaUser;

  module.exports.createSocialMediaUser = function(socialMediaUserData, callback){

    var cb3 = function(dontCare,user){
      var socialMediaUser = new SocialMediaUser({
        providerName: socialMediaUserData.providerName,
        providerUserId: socialMediaUserData.providerUserId,
        displayName: socialMediaUserData.displayName,
        user: user._id
      });

      socialMediaUser.save(function(err){
        if(err){
          process.nextTick(function(){
             callback(err);
          });
        }else{
          user.socialMediaPersonae.push(socialMediaUser._id);
          user.save(function(err){
            if(err){
              SocialMediaUser.remove({_id: socialMediaUser._id}, function(dontCare){});
              process.nextTick(function(){
                callback(err);
              });
            }else{
              process.nextTick(function(){
                findSocialMediaUser(socialMediaUser.providerName, socialMediaUser.providerUserId, callback);
              });
            }
          });
        }
      });
    };

    var cb2 = function(){
        var userObj = {
          firstName: socialMediaUserData.firstName,
            lastName: socialMediaUserData.lastName,
            userName: socialMediaUserData.userName,
            email: socialMediaUserData.email,
            password: "Random-Date-pwd_" + Math.floor(Math.random()*1000)
          };
          createNewUser(userObj, cb3);
      };

      var cb1 = function(err, user){
        if(err){
          process.nextTick(function(){
            callback(err);
          });
        }else if(!user){
          process.nextTick(function(){
            cb2();
          });
        }else{
          process.nextTick(function(){
            cb3(null, user);
          });
        }
      };

      findUserByEmail(socialMediaUserData.email, cb1);

    };

    /**
     * This function assumes all input has been validated beforehand
     * @param userData an object containing all the values to create the user
     * @param callback
     */
    function createNewUser(userData, callback) {
      var password = userData.password;
      bcryptUtils.generateSaltAndHash(password, function (err, saltAndHash) {
        if (err) {
          process.nextTick(function () {
            callback(err);
          });

        } else {
          var salt = saltAndHash.salt;
          var hash = saltAndHash.hash;

          var user = new User({
            firstName: userData.firstName,
            lastName: userData.lastName,
            userName: userData.userName,
            email: userData.email,
            salt: salt,
            password: hash,
            creationDate: new Date(),
            socialMediaPersonae: []
          });
          user.save(function (err) {
            if (err) {
              process.nextTick(function () {
                callback(err);
              });
            } else {
              process.nextTick(function () {
                callback(null, user);
              });
            }
          });
        }
      });
    }

    module.exports.createNewUser = createNewUser;
}).call(this);